audioutils/alsa-lib/alsa_pcm.c

```diff
@@ -0,0 +1,1209 @@
+{
+  if (format > SND_PCM_FORMAT_LAST)
+    {
+      return NULL;
+    }
+
+  return snd_pcm_format_names[format];
+}
+
+/****************************************************************************
+ * Hw Functions
+ ****************************************************************************/
+
+static inline int hw_is_mask(int var)
+{
+  return var >= SND_PCM_HW_PARAM_FIRST_MASK &&
+         var <= SND_PCM_HW_PARAM_LAST_MASK;
+}
+
+static inline int hw_is_range(int var)
+{
+  return var >= SND_PCM_HW_PARAM_FIRST_RANGE &&
+         var <= SND_PCM_HW_PARAM_LAST_RANGE;
+}
+
+static int _snd_pcm_hw_param_set(FAR snd_pcm_hw_params_t *params,
+                                 snd_pcm_hw_param_t var, unsigned int val)
+{
+  FAR union snd_interval *interval = NULL;
+
+  assert(params);
+  interval = &params->intervals[var - SND_PCM_HW_PARAM_FIRST_INTERVAL];
+  if (hw_is_mask(var))
+    {
+      interval->mask = (1 << val);
+    }
+  else
+    {
+      interval->range.min = val;
+    }
+
+  return 0;
+}
+
+static int _snd_pcm_hw_param_get(FAR const snd_pcm_hw_params_t *params,
+                                 snd_pcm_hw_param_t var,
+                                 FAR unsigned int *val)
+{
+  FAR const union snd_interval *interval;
+
+  assert(params);
+  interval = &params->intervals[var - SND_PCM_HW_PARAM_FIRST_INTERVAL];
+  if (hw_is_mask(var))
+    {
+      *val = (interval->mask >> 1);
+    }
+  else
+    {
+      *val = interval->range.min;
+    }
+
+  return 0;
+}
+
+#define _snd_pcm_hw_param_set(params, var, val)                           \
+  _snd_pcm_hw_param_set(params, var, (unsigned int)(val))
+#define _snd_pcm_hw_param_get(params, var, val)                           \
+  _snd_pcm_hw_param_get(params, var, (FAR unsigned int *)(val))
+
+static int _snd_pcm_hw_params_internal(FAR snd_pcm_t *pcm,
+                                       FAR snd_pcm_hw_params_t *params)
+{
+  int period_size;
+  int buffer_size;
+
+  assert(pcm && params);
+
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_FORMAT, &pcm->format);
+  if (!pcm->format)
+    {
+      pcm->format = SND_PCM_FORMAT_S16;
+    }
+
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_CHANNELS, &pcm->channels);
+  if (!pcm->channels)
+    {
+      pcm->channels = 2;
+    }
+
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_RATE, &pcm->sample_rate);
+  if (!pcm->sample_rate)
+    {
+      pcm->sample_rate = 44100;
+    }
+
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_PERIODS, &pcm->periods);
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_PERIOD_TIME,
+                        &pcm->period_time);
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_PERIOD_SIZE,
+                        &period_size);
+  _snd_pcm_hw_param_get(params, SND_PCM_HW_PARAM_BUFFER_SIZE,
+                        &buffer_size);
+  if (period_size && buffer_size)
+    {
+      pcm->periods = buffer_size / period_size;
+      pcm->period_time = period_size * 1000 * 1000 / pcm->sample_rate;
```

Comment:
should we save period_size, not period_time internally to avoid the conversion in multiple places